src/gui/editors/AutomationEditor.cpp

```diff
@@ -663,418 +637,175 @@ void AutomationEditor::mouseReleaseEvent(QMouseEvent * mouseEvent )
 	{
 		m_mouseDownRight = false;
 		mustRepaint = true;
 	}
 
-	if( mouseEvent->button() == Qt::LeftButton )
-	{
-		mustRepaint = true;
-	}
-
-	if( m_editMode == DRAW )
+	if (m_editMode == DRAW)
 	{
-		if( m_action == MOVE_VALUE )
+		if (m_action == MOVE_VALUE)
 		{
+			// Actually apply the value of the node being dragged
 			m_pattern->applyDragValue();
 		}
+
 		QApplication::restoreOverrideCursor();
 	}
 
 	m_action = NONE;
 
-	if( mustRepaint )
-	{
-		repaint();
-	}
-}
-
-
-
-
-void AutomationEditor::removePoints( int x0, int x1 )
-{
-	int deltax = qAbs( x1 - x0 );
-	int x = x0;
-	int xstep;
-
-	if( deltax < 1 )
-	{
-		return;
-	}
-
-	if( x0 < x1 )
-	{
-		xstep = 1;
-	}
-	else
-	{
-		xstep = -1;
-	}
-
-	int i = 0;
-	while( i <= deltax )
-	{
-		m_pattern->removeValue( TimePos( x ) );
-		x += xstep;
-		i += 1;
-	}
+	if (mustRepaint) { repaint(); }
 }
 
 
 
 
 void AutomationEditor::mouseMoveEvent(QMouseEvent * mouseEvent )
 {
-	QMutexLocker m( &m_patternMutex );
 	if( !validPattern() )
 	{
 		update();
 		return;
 	}
 
-	if( mouseEvent->y() > TOP_MARGIN )
+	// If the mouse y position is inside the Automation Editor viewport
+	if (mouseEvent->y() > TOP_MARGIN)
 	{
-		float level = getLevel( mouseEvent->y() );
-		int x = mouseEvent->x();
+		float level = getLevel(mouseEvent->y());
+		// Get the viewport X position where the mouse is at
+		int x = mouseEvent->x() - VALUES_WIDTH;
 
-		x -= VALUES_WIDTH;
-		if( m_action == MOVE_VALUE )
-		{
-			x -= m_moveXOffset;
-		}
+		// Get the X position in ticks
+		int posTicks = (x * TimePos::ticksPerBar() / m_ppb) + m_currentPosition;
 
-		int pos_ticks = x * TimePos::ticksPerBar() / m_ppb +
-							m_currentPosition;
-		// m_mouseDownLeft used to prevent drag when drawing line
-		if (m_mouseDownLeft && m_editMode == DRAW)
+		switch (m_editMode)
 		{
-			if( m_action == MOVE_VALUE )
+			case DRAW:
 			{
-				// moving value
-				if( pos_ticks < 0 )
+				// We are dragging a node
+				if (m_mouseDownLeft)
 				{
-					pos_ticks = 0;
+					if (m_action == MOVE_VALUE)
+					{
+						// When we clicked the node, we might have clicked slightly off
+						// so we account for that offset for a smooth drag
+						x -= m_moveXOffset;
+						posTicks = (x * TimePos::ticksPerBar() / m_ppb) + m_currentPosition;
+
+						// If we moved the mouse past the beginning correct the position in ticks
+						posTicks = qMax(posTicks, 0);
+
+						m_drawLastTick = posTicks;
+						m_drawLastLevel = level;
+
+						// Updates the drag value of the moved node
+						m_pattern->setDragValue(
+							TimePos(posTicks),
+							level,
+							true,
+							mouseEvent->modifiers() & Qt::ControlModifier
+						);
+
+						Engine::getSong()->setModified();
+					}
+					else if (m_action == DRAW_LINE)
+					{
+						// We are drawing a line. For now do nothing (as before), but later logic
+						// could be added here so the line is updated according to the new mouse position
+						// until the button is released
+					}
```

Comment:
This block will be removed during optimization pass, but let's still just comment out the `else if` since nothing is done.
```suggestion
					/* else if (m_action == DRAW_LINE)
					{
						// We are drawing a line. For now do nothing (as before), but later logic
						// could be added here so the line is updated according to the new mouse position
						// until the button is released
					}
					*/
```